## Scripts for RNASeq analysis
`Pipeline` .     
The following scripts are used:
`fastqc.sh`: quality control .   
`Trim_skewer.sh`: trim adaptors and low quality reads .      
`mapping_STAR_genome1.0.sh`: align reads to peach genome v1.0 .     
`mapping_STAR_genome2.0.sh`: align reads to peach genome v2.0 .      
`counting_HTSeq_genome1.0.sh`: count reads aligned to genome v1.0 .     
`counting_HTSeq_genome2.0.sh`: count reads aligned to genome v2.0 .     
`rRNA_filter_Bowtie2.sh`: filter out the rRNA reads .     
`merge_counts.sh`: create a table of read-counts for all samples .     



